staging/src/k8s.io/cloud-provider/controllers/service/controller.go

```diff
@@ -682,28 +715,83 @@ func (s *Controller) nodeSyncLoop() {
 	// round.
 	s.servicesToUpdate = s.cache.allServices()
 	numServices := len(s.servicesToUpdate)
-	s.servicesToUpdate = s.updateLoadBalancerHosts(s.servicesToUpdate, newHosts)
+	s.servicesToUpdate = s.updateLoadBalancerHosts(s.servicesToUpdate, workers)
 	klog.V(2).Infof("Successfully updated %d out of %d load balancers to direct traffic to the updated set of nodes",
 		numServices-len(s.servicesToUpdate), numServices)
 
 	s.knownHosts = newHosts
 }
 
-// updateLoadBalancerHosts updates all existing load balancers so that
-// they will match the list of hosts provided.
-// Returns the list of services that couldn't be updated.
-func (s *Controller) updateLoadBalancerHosts(services []*v1.Service, hosts []*v1.Node) (servicesToRetry []*v1.Service) {
-	for _, service := range services {
+type ResultContext struct {
+	service *v1.Service
+	retry   bool
+}
+
+func (s *Controller) nodeSyncWorker(svcCh chan *v1.Service, resultCh chan ResultContext) {
+	klog.V(4).Infof("nodeSyncWorker Started")
+	for svc := range svcCh {
 		func() {
-			if service == nil {
+			result := ResultContext{service: svc}
+			defer func() {
+				resultCh <- result
+			}()
+			if svc == nil {
 				return
 			}
-			if err := s.lockedUpdateLoadBalancerHosts(service, hosts); err != nil {
-				runtime.HandleError(fmt.Errorf("failed to update load balancer hosts for service %s/%s: %v", service.Namespace, service.Name, err))
-				servicesToRetry = append(servicesToRetry, service)
+			klog.V(4).Infof("nodeSyncWorker processing service %s/%s", svc.Namespace, svc.Name)
+			hosts, err := listWithPredicate(s.nodeLister, s.getNodeConditionPredicate())
+			if err != nil {
+				runtime.HandleError(fmt.Errorf("failed to retrieve node list: %v", err))
+				result.retry = true
+				return
+			}
+
+			if err := s.lockedUpdateLoadBalancerHosts(svc, hosts); err != nil {
+				runtime.HandleError(fmt.Errorf("failed to update load balancer hosts for service %s/%s: %v", svc.Namespace, svc.Name, err))
+				result.retry = true
+				return
 			}
 		}()
 	}
+	klog.V(4).Infof("nodeSyncWorker finished")
+}
+
+// updateLoadBalancerHosts updates all existing load balancers so that
+// they will match the list of hosts provided.
+// Returns the list of services that couldn't be updated.
+func (s *Controller) updateLoadBalancerHosts(services []*v1.Service, workers int) (servicesToRetry []*v1.Service) {
+	klog.V(4).Infof("Running updateLoadBalancerHosts(len(services)==%d, workers==%d)", len(services), workers)
+
+	if len(services) == 0 {
```

Comment:
You can simplify this function a lot by just doing something like:

```
lock := sync.Mutex{}

doWork := func(piece int) {
  if shouldRetry := nodeSyncService(services[i]); !shouldRetry {
    return
  }
  lock.Lock()
  defer lock.Unlock()
  servicesToRetry = append(servicesToRetry, services[i])
}
 
workqueue.ParallelizeUntil(context.TODO(), workers, doWork)
```

Reply:
Good call. Don't know this util exists.

Reply:
I tried this one out and notice that the performance gain is not as good as the current implementation. This is because the ParallelizeUntil preallocated the job pieces for each worker. However, some LB service may need a shorter time to update while others need more time. This may result in some worker finished first and the whole flow end up waiting for the worker which takes the longest.

An example is as follows: 
10 LB needs to get updated. (5 NetLBs takes longer, 5 ILBs takes shorter. ) 2 workers. 

Using the `ParallelizeUntil` function may end up allocating all 5 NetLBs for 1 worker making it while the other worker would finish early. 

However, using the current approach would make sure both workers are busy till the last one.

Reply:
I don't understand - ParallelizeUntil is doing exactly the same
(assuming you don't set chunkSize):

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/util/workqueue/parallelizer.go#L59
chunks = pieces
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/util/workqueue/parallelizer.go#L62
everything put into channel
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/util/workqueue/parallelizer.go#L79
take an item from channel
start = chunk
end = chunk+1
for start to end is exactly one item


This is exactly the same you have here.

So your description above isn't correct...

Reply:
oh my bad. I misunderstood it. You were right. I modify the code to use the util